src/libkomwm.py

```diff
@@ -224,21 +236,24 @@ def addPattern(dashes):
 
     # Parse style mapcss
     global style
-    style = MapCSS(options.minzoom, options.maxzoom + 1)
+    style = MapCSS(options.minzoom, options.maxzoom)
     style.parse(filename=options.filename, static_tags=mapcss_static_tags,
                 dynamic_tags=mapcss_dynamic_tags)
 
-    # Build optimization tree - class/type -> StyleChoosers
+    # Build optimization tree - class/zoom/type -> StyleChoosers
+    clname_cltag_unique = set()
     for cl in class_order:
         clname = cl if cl.find('-') == -1 else cl[:cl.find('-')]
         # Get first tag of the class/type.
         cltag = next(iter(classificator[cl].keys()))
-        style.build_choosers_tree(clname, "line", cltag)
-        style.build_choosers_tree(clname, "area", cltag)
-        style.build_choosers_tree(clname, "node", cltag)
-    style.restore_choosers_order("line")
-    style.restore_choosers_order("area")
-    style.restore_choosers_order("node")
+        clname_cltag = clname + '$' + cltag
```

Comment:
Could you please clarify here? If we have equal styles entry, we assume "overriding".
Or what do we _filter_ here?

Reply:
Sorry I didn't get what do you mean...
		Here we want to build the pre-match tree only for unique (type, zoom, clname),
		but there is a (only one) case where clname doesn't match the first tag:
		`building|address;[addr:housenumber][addr:street]`
		so the pre-match tree made for clname=="building" woulnd't have contained choosers applicable to `[addr:housenumber][addr:street]`, so I had to add the first tag into the unique mix too.